### PR TITLE
openpgl: Fix building on aarch64

### PR DESCRIPTION
**Pull request by Biswa96**

### CODE REVIEW

**Comment by Biswa96** [comment]
@hmartinez82 Would you like to check if this can be compiled in aarch64? Also if possible try to build blender with this. I have some doubt with the changes and want to ask experts about that.

**Comment by hmartinez82** [comment]
Success
` ` ` 
==>  MINGW_ARCH: clangarm64
  -> Building clangarm64...
==> Making package: mingw-w64-openpgl 0.4.1-3 (Sun, Mar 26, 2023  3:06:44 AM)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> Retrieving sources...
  -> Found openpgl-0.4.1.tar.gz
  -> Found 001-add-compile-flags.patch
  -> Found 002-Fix-finding-TBB.patch
  -> Found 003-Fix-building-on-mingw-w64-aarch64.patch
==> Validating source files with sha256sums...
    openpgl-0.4.1.tar.gz ... Passed
    001-add-compile-flags.patch ... Passed
    002-Fix-finding-TBB.patch ... Passed
    003-Fix-building-on-mingw-w64-aarch64.patch ... Passed
==> Removing existing $srcdir/ directory...
==> Extracting sources...
  -> Extracting openpgl-0.4.1.tar.gz with bsdtar
==> Starting prepare()...
patching file openpgl/CMakeLists.txt
patching file openpgl/CMakeLists.txt
patching file openpgl/openpgl_common.h
patching file third-party/embreeSrc/common/simd/arm/sse2neon.h
patching file third-party/embreeSrc/common/sys/intrinsics.h
==> Removing existing $pkgdir/ directory...
==> Starting build()...
-- The CXX compiler identification is Clang 15.0.7
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/msys64/clangarm64/bin/clang++.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The C compiler identification is Clang 15.0.7
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/msys64/clangarm64/bin/clang.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- CMAKE_CURRENT_LIST_DIR = C:/M_P/mingw-w64-openpgl/src/openpgl-0.4.1-beta/openpgl
-- CMAKE_CURRENT_SOURCE_DIR = C:/M_P/mingw-w64-openpgl/src/openpgl-0.4.1-beta/openpgl
-- CMAKE_CURRENT_BINARY_DIR = C:/M_P/mingw-w64-openpgl/src/build-CLANGARM64/openpgl
-- Configuring done (1.7s)
-- Generating done (0.0s)
-- Build files have been written to: C:/M_P/mingw-w64-openpgl/src/build-CLANGARM64
[8/8] Linking CXX shared library openpgl\libopenpgl.dll
==> Starting package()...
-- Install configuration: "Release"
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/lib/cmake/openpgl-0.4.1/openpglConfig.cmake
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/lib/cmake/openpgl-0.4.1/openpglConfigVersion.cmake
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/lib/libopenpgl.dll.a
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/bin/libopenpgl.dll
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/lib/cmake/openpgl-0.4.1/openpgl_Exports.cmake
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/lib/cmake/openpgl-0.4.1/openpgl_Exports-release.cmake
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/common.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/config.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/cpp
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/cpp/Common.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/cpp/Device.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/cpp/Distribution.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/cpp/Field.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/cpp/OpenPGL.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/cpp/PathSegment.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/cpp/PathSegmentStorage.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/cpp/Region.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/cpp/SampleData.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/cpp/Sampler.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/cpp/SampleStorage.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/cpp/SurfaceSamplingDistribution.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/cpp/VolumeSamplingDistribution.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/data.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/device.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/distribution.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/field.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/openpgl.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/pathsegmentstorage.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/region.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/sampler.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/samplestorage.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/surfacesamplingdistribution.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/types.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/volumesamplingdistribution.h
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/share/doc/openpgl/LICENSE.txt
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/share/doc/openpgl/third-party-programs.txt
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/share/doc/openpgl/third-party-programs-Embree.txt
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/share/doc/openpgl/third-party-programs-TBB.txt
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/share/doc/openpgl/README.md
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/share/doc/openpgl/CHANGELOG.md
-- Installing: C:/M_P/mingw-w64-openpgl/pkg/mingw-w64-clang-aarch64-openpgl/clangarm64/include/openpgl/version.h
==> Tidying install...
  -> Removing libtool files...
  -> Purging unwanted files...
  -> Stripping unneeded symbols from binaries and libraries...
  -> Compressing man and info pages...
==> Checking for packaging issues...
==> Creating package "mingw-w64-clang-aarch64-openpgl"...
  -> Generating .PKGINFO file...
  -> Generating .BUILDINFO file...
  -> Generating .MTREE file...
  -> Compressing package...
==> Finished making: mingw-w64-openpgl 0.4.1-3 (Sun, Mar 26, 2023  3:07:12 AM)
 ` ` `

File: mingw-w64-openpgl/003-Fix-building-on-mingw-w64-aarch64.patch [hunk 1]
```diff
@@ -31,3 +44,38 @@
...
+     size_t i = bsf(v);
+@@ -138,9 +138,9 @@
+ #endif
+   }
+   
+-#if defined(__X86_64__)
++#if defined(__X86_64__) || defined (__aarch64__)
+   __forceinline size_t bsr(size_t v) {
+-#if defined(__AVX2__) 
++#if defined(__AVX2__) && !defined(__aarch64__)
```
Comment (lazka) [inline mingw-w64-openpgl/003-Fix-building-on-mingw-w64-aarch64.patch:78]: isn't AVX2 a x86 only thing?

**Comment by Biswa96** [comment]
Is it possible to check if openpgl can be used with blender now? I am little doubtful about the `#include <mm_malloc.h>` addition. It is added for x86 and x86_64 in mingw-w64 but not for aarch64.

File: mingw-w64-openpgl/003-Fix-building-on-mingw-w64-aarch64.patch (hunk 1 above)
Comment (Biswa96) [reply mingw-w64-openpgl/003-Fix-building-on-mingw-w64-aarch64.patch:78]: Yes, it is required. I am not sure about the reason. The openpgl/CMakeLists.txt file defines `__AVX2__` if neon is present. Maybe it is required for sse2neon translation.

Comment (Biswa96) [reply mingw-w64-openpgl/003-Fix-building-on-mingw-w64-aarch64.patch:78]: BTW, if you see the `int bsr(int v)` function, it also has that `!defined(__aarch64__)` condition.

Comment (lazka) [reply mingw-w64-openpgl/003-Fix-building-on-mingw-w64-aarch64.patch:78]: ok, I see. confusing...

**Comment by MehdiChinoune** [comment]
There is another file avx2neon.h

**Comment by MehdiChinoune** [comment]
Could it be merged?

**Pull request merged**